File: 33_search_in_rotated_sorted_array/search_in_rotated_sorted_array.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        k = 0
        for i in range(1, len(nums)):
            if nums[i - 1] > nums[i]:
                k = i
                break
        nums = nums[k:] + nums[:k]
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                result = mid + k
                if result > len(nums) - 1:
                    result %= len(nums)
                return result
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        return -1

File: 33_search_in_rotated_sorted_array/search_in_rotated_sorted_array.java
class Solution {
    public int search(int[] nums, int target) {
        // 直接二分法 即使数组旋转后不是有序 但二分成两部分时 一定有一部分是有序的
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            }
            if(nums[0] <= nums[mid]) {
                if (nums[0] <= target && target < nums[mid]) {
                    right = mid -1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[nums.length - 1]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
}